File: Java/Sample_Source/Text/RegexExtractText/src/com/datalogics/PDFL/Samples/RegexExtractText.java
package com.datalogics.pdfl.samples.Text.RegexExtractText;

import com.datalogics.PDFL.DocTextFinder;
import com.datalogics.PDFL.DocTextFinderMatch;
import com.datalogics.PDFL.DocTextFinderQuadInfo;
import com.datalogics.PDFL.Document;
import com.datalogics.PDFL.Library;
import com.datalogics.PDFL.Page;
import com.datalogics.PDFL.Point;
import com.datalogics.PDFL.Quad;
import com.datalogics.PDFL.SaveFlags;
import com.datalogics.PDFL.WordFinderConfig;

import java.util.EnumSet;
import java.io.FileWriter;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONObject;

/*
 * 
 * This sample demonstrates using DocTextFinder to find instances of a phrase
 * that matches a user-supplied regular expression. The output is a JSON file that
 * has the match information.
 *
 * Copyright (c) 2021, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */

// This Datalogics sample uses the org.json (JSON-Java) library to generate JSON output. Below is the JSON license for the org.json software:
/*
Copyright (c) 2002 JSON.org
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
The Software shall be used for Good, not Evil.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.
*/

public class RegexExtractText {
	public static void main (String[] args) throws Throwable
    {
        System.out.println("RegexExtractText sample:");

        Library lib = new Library();

        try {

            final String sInput = (args.length > 0) ? args[0] : Library.getResourceDirectory() + "Sample_Input/RegexExtractText.pdf";
            final String sOutput = "RegexExtractText-out.json";

            // Phone numbers
            final String sRegex = "((1-)?(\\()?\\d{3}(\\))?(\\s)?(-)?\\d{3}-\\d{4})";
            // Email addresses
            //String sRegex = "(\\b[\\w.!#$%&'*+\\/=?^`{|}~-]+@[\\w-]+(?:\\.[\\w-]+)*\\b)";
            // URLs
            //String sRegex = "((https?:\\/\\/(?:www\\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\\.[^\\s]{2,}|www\\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\\.[^\\s]{2,}|https?:\\/\\/(?:www\\.|(?!www))[a-zA-Z0-9]+\\.[^\\s]{2,}|www\\.[a-zA-Z0-9]+\\.[^\\s]{2,}))";
            
            System.out.println("Reading " + sInput);

            Document doc = new Document(sInput);

            // This array will hold the JSON stream that we will print to the output JSON file.
            JSONArray result = new JSONArray();
            
            final int nPages = doc.getNumPages();

            System.out.println("Opened document " + sInput);
            
            WordFinderConfig wordConfig = new WordFinderConfig();
            
            // Need to set this to true so phrases will be concatenated properly.
            wordConfig.setNoHyphenDetection(true);
            
            DocTextFinder docTextFinder = new DocTextFinder(doc, wordConfig);
            
            List<DocTextFinderMatch> docMatches = docTextFinder.getMatchList(0, nPages - 1, sRegex);

            for (DocTextFinderMatch wInfo : docMatches) {
                // This JSON object will store the match phrase and an array of quads for the match.
                JSONObject matchObject = new JSONObject();

                // This JSON array will store the page number and quad location for each match quad.
                JSONArray matchQuadInformation = new JSONArray();

                // Set the match phrase in the JSON object.
                matchObject.put("match-phrase", wInfo.getMatchString());

                // Get the word quads
                List<DocTextFinderQuadInfo> quadList = wInfo.getQuadInfo();

                // Iterate through the quad info
                for (DocTextFinderQuadInfo qInfo : quadList) {
                    for(Quad quad : qInfo.getQuads()){
                        // Get the coordinates of the quad.
                        double top_left_x = quad.getTopLeft().getH();
                        double top_left_y = quad.getTopLeft().getV();

                        double bottom_left_x = quad.getBottomLeft().getH();
                        double bottom_left_y = quad.getBottomLeft().getV();

                        double top_right_x = quad.getTopRight().getH();
                        double top_right_y = quad.getTopRight().getV();

                        double bottom_right_x = quad.getBottomRight().getH();
                        double bottom_right_y = quad.getBottomRight().getV();

                        // Set the quad coordinates in JSON objects.
                        JSONObject topLeft = new JSONObject();
                        topLeft.put("x", top_left_x);
                        topLeft.put("y", top_left_y);
                        
                        JSONObject bottomLeft = new JSONObject();
                        bottomLeft.put("x", bottom_left_x);
                        bottomLeft.put("y", bottom_left_y);
                        
                        JSONObject topRight = new JSONObject();
                        topRight.put("x", top_right_x);
                        topRight.put("y", top_right_y);
                        
                        JSONObject bottomRight = new JSONObject();
                        bottomRight.put("x", bottom_right_x);
                        bottomRight.put("y", bottom_right_y);

                        // Use the quad coordinate JSON objects to form a single JSON object that holds match quad location information.
                        JSONObject quadLocation = new JSONObject();
                        quadLocation.put("bottom-left", bottomLeft);
                        quadLocation.put("bottom-right", bottomRight);
                        quadLocation.put("top-left", topLeft);
                        quadLocation.put("top-right", topRight);

                        JSONObject quadInformationObject = new JSONObject();
                        quadInformationObject.put("page-number", qInfo.getPageNum());
                        quadInformationObject.put("quad-location", quadLocation);

                        // Insert the match's page number and quad location(s) in the matchQuadInformation JSON array.
                        matchQuadInformation.put(quadInformationObject);
                    }
                }
                // Set the match's quad information in the matchObject.
                matchObject.put("match-quads", matchQuadInformation);

                result.put(matchObject);
            }
            // Write the match information to the output JSON file.
            System.out.println("Writing JSON to " + sOutput);
            FileWriter file = new FileWriter(sOutput);
            file.write(result.toString(4));
            file.close();

            doc.close();
        }
        finally {
            lib.delete();
        }
    }
}

File: DotNETFramework/Sample_Source/Text/RegexExtractText/RegexExtractText.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Newtonsoft.Json;
using Datalogics.PDFL;

/*
 * This sample demonstrates using DocTextFinder to find instances of a phrase
 * that matches a user-supplied regular expression. The output is a JSON file that
 * has the match information.
 * 
 * Copyright (c) 2021, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */

// This Datalogics sample uses the Newtonsoft.Json library to generate JSON output. Below is the MIT license for the Newtonsoft.Json software:
/*
Copyright (c) 2007 James Newton-King

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

namespace RegexExtractText
{
    // This class represents a match quad's bottom left coordinate.
    public class BottomLeft {
        public double x { get; set; }
        public double y { get; set; }
    }

    // This class represents a match quad's bottom right coordinate.
    public class BottomRight {
        public double x { get; set; }
        public double y { get; set; }
    }

    // This class represents a match quad's top left coordinate.
    public class TopLeft {
        public double x { get; set; }
        public double y { get; set; }
    }

    // This class represents a match quad's top right coordinate.
    public class TopRight {
        public double x { get; set; }
        public double y { get; set; }
    }

    // This class represents the 4 coordinates that make up a match quad.
    public class QuadLocation {
        [JsonProperty("bottom-left")]
        public BottomLeft bottomLeft { get; set; }

        [JsonProperty("bottom-right")]
        public BottomRight bottomRight { get; set; }

        [JsonProperty("top-left")]
        public TopLeft topLeft { get; set; }

        [JsonProperty("top-right")]
        public TopRight topRight { get; set; }
    }

    // This class represents a match quad's location (the quad coordinates and page number that quad is located on).
    public class MatchQuadInformation {
        [JsonProperty("page-number")]
        public int pageNumber { get; set; }

        [JsonProperty("quad-location")]
        public QuadLocation quadLocation { get; set; }
    }

    // This class represents the information that is associated with a match (match phrase and match quads).
    public class MatchObject {
        [JsonProperty("match-phrase")]
        public string matchPhrase { get; set; }

        [JsonProperty("match-quads")]
        public List<MatchQuadInformation> matchQuads { get; set; }
    }

    // This class represents the final JSON that will be written to the output JSON file.
    public class DocTextFinderJson {
        public List<MatchObject> documentJson;
    }

    class RegexExtractText
    {
        static void Main(string[] args)
        {
            Console.WriteLine("RegexExtractText Sample:");

            using (Library lib = new Library())
            {
                Console.WriteLine("Initialized the library.");

                String sInput = Library.ResourceDirectory + "Sample_Input/RegexExtractText.pdf";
                String sOutput = "../RegexExtractText-out.json";

                // Uncomment only one regular expression you are interested in seeing the match information of (as a JSON file).
                // Phone numbers
                String sRegex = "((1-)?(\\()?\\d{3}(\\))?(\\s)?(-)?\\d{3}-\\d{4})";
                // Email addresses
                //String sRegex = "(\\b[\\w.!#$%&'*+\\/=?^`{|}~-]+@[\\w-]+(?:\\.[\\w-]+)*\\b)";
                // URLs
                //String sRegex = "((https?:\\/\\/(?:www\\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\\.[^\\s]{2,}|www\\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\\.[^\\s]{2,}|https?:\\/\\/(?:www\\.|(?!www))[a-zA-Z0-9]+\\.[^\\s]{2,}|www\\.[a-zA-Z0-9]+\\.[^\\s]{2,}))";

                if (args.Length > 0)
                    sInput = args[0];

                using (Document doc = new Document(sInput))
                {
                    int nPages = doc.NumPages;

                    Console.WriteLine("Input file:  " + sInput);

                    // This will hold the JSON stream that we will print to the output JSON file.
                    DocTextFinderJson result = new DocTextFinderJson();
                    result.documentJson = new List<MatchObject>();

                    WordFinderConfig wordConfig = new WordFinderConfig();

                    // Need to set this to true so phrases will be concatenated properly.
                    wordConfig.NoHyphenDetection = true;

                    // Create a DocTextFinder with the default wordfinder parameters.
                    using (DocTextFinder docTextFinder =
                        new DocTextFinder(doc, wordConfig))
                    {
                        // Retrieve the phrases matching a regular expression.
                        IList<DocTextFinderMatch> docMatches =
                            docTextFinder.GetMatchList(0, nPages - 1, sRegex);

                        // Iterate through the matches and add match information to the DocTextFinderJson object.
                        foreach (DocTextFinderMatch wInfo in docMatches)
                        {
                            // This object will store the match phrase and an array of quads for the match.
                            MatchObject matchObject = new MatchObject();
                            
                            // This list will store the page number and quad location for each match quad.
                            List<MatchQuadInformation> matchQuadInformationList = new List<MatchQuadInformation>();

                            // Set the match phrase in the matchObject.
                            matchObject.matchPhrase = wInfo.MatchString;

                            // Get the quads.
                            IList<DocTextFinderQuadInfo> QuadInfo = wInfo.QuadInfo;

                            foreach (DocTextFinderQuadInfo qInfo in QuadInfo)
                            {
                                MatchQuadInformation temp = new MatchQuadInformation();
                                temp.pageNumber = qInfo.PageNum;

                                // Iterate through the quads and insert the quad information into the matchQuadInformation object.
                                foreach (Quad quad in qInfo.Quads)
                                {
                                    QuadLocation quadLocation = new QuadLocation();
                                    quadLocation.topLeft = new TopLeft();
                                    quadLocation.bottomLeft = new BottomLeft();
                                    quadLocation.topRight = new TopRight();
                                    quadLocation.bottomRight = new BottomRight();

                                    quadLocation.topLeft.x = quad.TopLeft.H;
                                    quadLocation.topLeft.y = quad.TopLeft.V;

                                    quadLocation.bottomLeft.x = quad.BottomLeft.H;
                                    quadLocation.bottomLeft.y = quad.BottomLeft.V;

                                    quadLocation.topRight.x = quad.TopRight.H;
                                    quadLocation.topRight.y = quad.TopRight.V;

                                    quadLocation.bottomRight.x = quad.BottomRight.H;
                                    quadLocation.bottomRight.y = quad.BottomRight.V;

                                    temp.quadLocation = quadLocation;
                                    matchQuadInformationList.Add(temp);
                                }
                            }
                            matchObject.matchQuads = matchQuadInformationList;
                            result.documentJson.Add(matchObject);
                        }
                        // Save the output JSON file.
                        Console.WriteLine("Writing JSON to " + sOutput);
                        string json = JsonConvert.SerializeObject(result.documentJson, Formatting.Indented);
                        System.IO.File.WriteAllText(sOutput, json);
                    }
                }
            }
        }
    }
}
